File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int>ans;
        int row = matrix.size(),col=matrix[0].size();
        int top=0,bottom=row-1,left=0,right=col-1;
        while(left<=right && top<=bottom){
            //print top

            for(int j=left;j<=right;j++){
                ans.push_back(matrix[top][j]);}
                top++;
            
            //print right
            for(int i=top;i<=bottom;i++){
                ans.push_back(matrix[i][right]);}
                right--;
            
            //print bottom
            if(top<=bottom){
            for(int j=right;j>=left;j--){
                ans.push_back(matrix[bottom][j]);}
                bottom--;
            }
            //print left
            if(left<=right){
            for(int i=bottom;i>=top;i--){
                ans.push_back(matrix[i][left]);}
                left++;
            }
        }return ans;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] arr) {
        int row=arr.length;
        int col=arr[0].length;
        int toprow=0;
        int bottomrow=row-1;
        int leftcol=0;
        int rightcol=col-1;
        ArrayList<Integer> ans=new ArrayList<>();
        while(toprow<=bottomrow && leftcol<=rightcol){
            for(int j=leftcol;j<=rightcol;j++){
                ans.add(arr[toprow][j]);
            }
            toprow++;
            for(int i=toprow;i<=bottomrow;i++){
                ans.add(arr[i][rightcol]);
            }
            rightcol--;
            if(toprow<=bottomrow){
                for(int j=rightcol;j>=leftcol;j--){
                    ans.add(arr[bottomrow][j]);
                }
                bottomrow--;
            }
            if(leftcol<=rightcol){
                for(int i=bottomrow;i>=toprow;i--){
                    ans.add(arr[i][leftcol]);
                }
                leftcol++;
            }
        }
        return ans;
    }
}